sources/slack/__init__.py

```diff
@@ -0,0 +1,132 @@
+"""Fetches Slack Conversations, History and logs."""
+
+from typing import Any, Dict, Iterable, List, Optional
+
+import dlt
+from dlt.common.typing import TAnyDateTime, TDataItem
+from dlt.extract.source import DltResource
+from pendulum import DateTime
+
+from .helpers import SlackAPI, ensure_dt_type, extract_jsonpath
+from .settings import DEFAULT_START_DATE, MAX_PAGE_SIZE
+
+
+@dlt.source(name="slack", max_table_nesting=2)
+def slack_source(
+    page_size: int = MAX_PAGE_SIZE,
+    access_token: str = dlt.secrets.value,
+    start_date: Optional[TAnyDateTime] = DEFAULT_START_DATE,
+    end_date: Optional[TAnyDateTime] = None,
+    selected_channels: Optional[List[str]] = dlt.config.value,
+) -> Iterable[DltResource]:
+    """
+    The source for the Slack pipeline. Available resources are conversations, conversations_history
+    and access_logs.
+
+    Args:
+        page_size: The max number of items to fetch per page. Defaults to 1000.
+        access_token: the oauth access_token used to authenticate.
+        start_date: The start time of the range for which to load. Defaults to January 1st 2000.
+        end_date: The end time of the range for which to load data.
+        selected_channels: The list of channels to load. If None, all channels will be loaded.
+
+    Returns:
+        Iterable[DltResource]: A list of DltResource objects representing the data resources.
+    """
+
+    end_dt: Optional[DateTime] = ensure_dt_type(end_date)
+    start_dt: Optional[DateTime] = ensure_dt_type(start_date)
+
+    api = SlackAPI(
+        access_token=access_token,
+        page_size=page_size,
+    )
+
+    channels: List[TDataItem] = []
+    for page_data in api.get_pages(
+        resource="conversations.list",
+        response_path="$.channels[*]",
+        datetime_fields=["updated", "created"],
+    ):
+        channels.extend(page_data)
+
+    @dlt.resource(name="channels", primary_key="id", write_disposition="replace")
+    def channels_resource() -> Iterable[TDataItem]:
+        """Yield all channels as a DLT resource."""
+        yield from channels
```

Comment:
why don't you put pagination code under the resource `channels_resource` function and don't use `add_filter` for selected_channels?

Reply:
then you can use channels with transformer as earlier. why did you changed it?